### ISSUE TITLE
Generalize config lookup mechanism

### ISSUE THREAD

**Issue by masaeedu** [editorconfig/editorconfig#408]
The editorconfig specification states:

> When opening a file, EditorConfig plugins look for a file named .editorconfig in the directory of the opened file and in every parent directory.

This assumes the source code we want to edit:

- a) is actually a file on a filesystem (as opposed to just being an arbitrary named buffer)
- b) is situated in a file tree that is well aligned with the desired organization of `.editorconfig`s

The second point bears a little explanation. Suppose we want to have separate, reusable `.editorconfig`s for JS, YAML, and NPM packages. We want to layer these in the order {JS, YAML, NPM}, with successive editorconfigs taking precedence over earlier ones.

In order to use these in a project, we somehow need to stuff these `.editorconfig`s in the correct order into the superfolder hierarchy of all projects for which they are applicable. This is not always convenient or possible.

It would be nice if we could decouple the organization of our `.editorconfig`s from the organization of our file system. I don't know what the right solution is exactly, but perhaps we could have an env var like:

```
EDITORCONFIGS='{ "/some/absolute/path": ["/path/to/editorconfig", "/path/to/another/editorconfig"], "/some/other/absolute/path": ["/another/editorconfig"] }'
```

The tools built around the editorconfig spec could simply take this mapping from folders to ordered lists of applicable editorconfigs as gospel whenever it is provided, and otherwise fall back to constructing the analogous mapping by inspecting the filesystem.

Allowing this kind of escape hatch would solve issues such as:

- https://github.com/editorconfig/editorconfig-vscode/issues/50
- https://github.com/editorconfig/editorconfig-vim/issues/54
- https://github.com/editorconfig/editorconfig/issues/236

**Comment by cxw42** [comment]
> Suppose we want to have separate, reusable .editorconfigs for JS, YAML, and NPM packages.

Would you please expand on this point?  I am not sure I understand the use case fully.  Are you developing NPM packages that include both js and yaml files?

**Comment by masaeedu** [comment]
@cxw42 

> Are you developing NPM packages that include both js and yaml files?

Yes. Sometimes I have even more exotic combinations, like PowerShell in a Haskell codebase, or Dhall in a Python codebase :). I'd rather not copy paste my desired settings for each of these languages into a bespoke editorconfig file for every project where they are applicable. I'd instead prefer to have these reusable "topical" `.editorconfig`s in a git repo, so that I can check them out into my home folder or adjacent to the projects that I want to work on, and then somehow inform the editors I'm using that these global editorconfigs apply to the projects I'm working on.

> They were even accepted into the main repo at $work!

Assuming you have many different repos, how do you manage company wide styles for specific languages/files? Are the settings for each language just copy/pasted into each repo where the language is being used?

**Comment by cxw42** [comment]
@masaeedu Thanks for the additional detail!  

Yes, in my use case I copy the settings for each language into each project.  I am indeed using bespoke per-project `.editorconfig` files.  So far I am the only heavy EditorConfig user at $work, so the question of company-wide styles hasn't yet arisen :D .

I see your point that there is merit to DRY.  I am not in charge here :) so defer to the Board, but I suspect this may be out of scope for EditorConfig.  The goal of EditorConfig (as I understand it) is to provide consistency on multi-developer projects.  Basing settings on files outside a project makes it harder to achieve that goal.

As an alternative, I would be happy to write or upvote a script that created a .editorconfig file by pulling from multiple sources.  Perhaps a workflow like:

In `.bashrc`:

    export EDITORCONFIG_PARTIALS=~/ecpartials

When starting a new project:

    cd Project_Root
    editorconfig-init yaml haskell perl
    # Now .editorconfig has the settings from ~/ecpartials/{yaml,haskell,perl}.editorconfig
    git add .editorconfig 

What do you think?

**Comment by masaeedu** [comment]
I don't think this solves the problem. This would essentially be a code generator, and since many nice code generators and scaffolding tools exist apart from editorconfig, an editorconfig-specific code generator wouldn't provide much value. If a user wants to change any of these settings (or add a new language-specific policy to apply to all codebases) you would still need to go around modifying these files in every repo.

Regarding:

> The goal of EditorConfig (as I understand it) is to provide consistency on multi-developer projects.

Yes, but I don't understand how what I'm suggesting is inconsistent with this. The lack of the feature described (i.e. the workflow of copy-pasting things between projects) actually makes it more difficult to maintain consistency in a multi-developer environment. If developers can point to a repository full of company-blessed configurations for every language (whether it is stored in their home folder, an adjacent repo, or even a submodule of the project), the code produced throughout the organization would look consistent regardless of what repository it's in, or who worked on it.

**Comment by cxw42** [comment]
Ah!  Seems like we're thinking about different use cases.  It's certainly technically possible.  As I said above, I defer to the Board.

**Comment by zenaan** [comment]
Here's a proof of concept for a syntax_uid map: https://github.com/zenaan/lang-list

**Comment by cxw42** [comment]
@zenaan Thanks - I'll take a look!
